### PR TITLE
[SYCL] Check CL_TARGET_OPENCL_VERSION at run time

### PR DESCRIPTION
**Pull request by KarachunIvan**
Replace compile time CL_TARGET_OPENCL_VERSION check with runtime.

Signed-off-by: Ivan Karachun <ivan.karachun@intel.com>

### CODE REVIEW

File: sycl/source/detail/sampler_impl.cpp [hunk 1]
```diff
@@ -46,34 +46,40 @@ RT::PiSampler sampler_impl::getOrCreateSampler(const context &Context) {
...
+        getSyclObjImpl(Context)->getHandleRef(), sprops, &errcode_ret),
+        errcode_ret));
+
+    if (errcode_ret == PI_INVALID_OPERATION)
+        throw feature_not_supported("Images are not supported by this device.");
+  } else {
+    // TODO: do we really need this old interface into PI and here?
+    cl_int cl_errcode_ret;
+    m_contextToSampler[Context] = pi::pi_cast<RT::PiSampler>(
+        clCreateSampler(pi::pi_cast<cl_context>(getSyclObjImpl(Context)->getHandleRef()),
```
Comment (KarachunIvan) [inline sycl/source/detail/sampler_impl.cpp]: @smaslov-intel, Is it OK that I call OpenCL 1.0 API like that? In current implementation PI always calls OpenCL 2.0 API so I have to call OpenCL 1.0 API directly. But the SYCL compiler is supposed to work with all target device runtimes regardless their version. I guess it's up to PI to decide which API it should call (I suggest to always call 1.0 API).

Comment (smaslov-intel) [reply sycl/source/detail/sampler_impl.cpp]: We should NOT call OpenCL directly ever, except where running OpenCL interop (and even that better go through PI). In this case (where different OpenCL versions have different interfaces) the  PI should be a unified interface that different plugins would be able to implement. To be more specific, if we really need to support OpenCL 1.0, then this if/else should go to the implementation of the piSamplerCreate in OpenCL plugin for PI (now in pi_opencl.cpp), and SYCL RT would just call piSamplerCreate always. OpenCL 2.0 version had already been chosen as an interface, and I don't see a reason to change that now.

**Comment by bader** [comment]
> OpenCL 2.0 version had already been chosen as an interface, and I don't see a reason to change that now.

The only reason to for that is to silence warnings coming from OpenCL Headers.
We shouldn't really use OpenCL 2.0 API for standard SYCL 1.2.1 features.
OpenCL 2.0+ API can be required only for extensions (e.g. USM).

File: sycl/source/detail/sampler_impl.cpp (hunk 1 above)
Comment (agozillon) [reply sycl/source/detail/sampler_impl.cpp]: I think a good existing example of this is perhaps OpenCL PI's piPlatformsGet (correct me if i am wrong though): https://github.com/intel/llvm/blob/sycl/sycl/source/detail/pi_opencl.cpp#L81 

I also plan to pull request in something similar for piQueueCreate in the near future.

Comment (smaslov-intel) [reply sycl/source/detail/sampler_impl.cpp]: Yes, exactly. Thanks for pointing this out. @KarachunIvan, please rework the sampler creation accordingly.

File: sycl/source/detail/pi_opencl.cpp [hunk 2]
```diff
@@ -159,6 +159,33 @@ pi_program OCL(piProgramCreate)(pi_context context, const void *il,
   return pi_cast<pi_program>(resProgram);
 }
 
+pi_sampler OCL(piSamplerCreate)(pi_context context,
+                                const cl_sampler_properties *sampler_properties,
+                                pi_result *error_code) {
+  cl_sampler returnSampler;
+  cl_bool normalizedCoords;
+  cl_addressing_mode addressingMode;
+  cl_filter_mode filterMode;
```
Comment (KarachunIvan) [inline sycl/source/detail/pi_opencl.cpp]: Should I initialize these parameters some default values?

File: sycl/source/detail/pi_opencl.cpp [hunk 3]
```diff
@@ -159,6 +159,33 @@ pi_program OCL(piProgramCreate)(pi_context context, const void *il,
...
+  // Unpack sampler properties
+  for (std::size_t i = 0; sampler_properties && sampler_properties[i] != 0;
+       ++i) {
+    if (sampler_properties[i] == CL_SAMPLER_NORMALIZED_COORDS) {
+      normalizedCoords = sampler_properties[++i];
+    } else if (sampler_properties[i] == CL_SAMPLER_ADDRESSING_MODE) {
+      addressingMode = sampler_properties[++i];
+    } else if (sampler_properties[i] == CL_SAMPLER_FILTER_MODE) {
+      filterMode = sampler_properties[++i];
+    }
```
Comment (KarachunIvan) [inline sycl/source/detail/pi_opencl.cpp]: Should I add another `else` branch that indicates that property name wasn't found?

File: sycl/source/detail/pi_opencl.cpp (hunk 2 above)
Comment (smaslov-intel) [reply sycl/source/detail/pi_opencl.cpp]: either this, or make sure they are all initialized below

File: sycl/source/detail/pi_opencl.cpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/source/detail/pi_opencl.cpp]: sure, that would be more robust. you can probably use PI_ASSERT

**Pull request merged**